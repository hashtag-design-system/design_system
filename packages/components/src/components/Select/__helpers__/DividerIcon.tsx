import React from "react";

export const DividerIcon: React.FC = () => {
  return (
    <svg width={2} className="select__input__divider" height={34} fill="none">
      <path d="M1 32V1" strokeWidth={1.5} strokeLinecap="round" />
    </svg>
  );
};
