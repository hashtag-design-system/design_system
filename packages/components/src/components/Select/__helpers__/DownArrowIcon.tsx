import { motion } from "framer-motion";
import React, { useContext } from "react";
import DropdownContext from "../../../utils/ctx/DropdownContext";

export const DownArrowIcon: React.FunctionComponent = () => {
  const selectCtx = useContext(DropdownContext);

  const { isVisible, handleClick } = selectCtx;

  // https://blog.lftechnology.com/using-svg-icons-components-in-react-44fbe8e5f91
  return (
    <motion.svg
      fill="none"
      viewBox="0 0 24 24"
      className="input__icon select__icon"
      // * A "spring" animation might be too much and useless for this element
      animate={isVisible ? { rotate: 180 } : { rotate: 0 }}
      onClick={e => handleClick && handleClick(e)}
    >
      <path d="M21 7l-9 10L3 7" strokeWidth={1.5} strokeLinecap="round" strokeLinejoin="round" />
    </motion.svg>
  );
};
