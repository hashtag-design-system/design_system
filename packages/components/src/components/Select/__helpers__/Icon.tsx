import { motion, SVGMotionProps } from "framer-motion";
import React from "react";

export const Icon: React.FC<{ clicked: boolean; onClick?: SVGMotionProps<SVGElement>["onClick"] }> = ({ clicked, ...props }) => {
  return (
    <motion.svg
      width={24}
      height={24}
      fill="none"
      className="input__icon select__icon"
      animate={clicked ? { rotate: 180 } : { rotate: 0 }}
      // * A "spring" animation might be too much and useless for this element
      {...props}
    >
      <path d="M21 7l-9 10L3 7" stroke="var(--grey-7)" strokeWidth={2} strokeLinecap="round" strokeLinejoin="round" />
    </motion.svg>
  );
};

export default Icon;
