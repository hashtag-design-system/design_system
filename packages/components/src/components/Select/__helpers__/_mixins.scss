@mixin select__item--hover() {
  outline: none;

  &:not(.select__header) {
    background-color: $grey-4;
    cursor: pointer;
  }
}

.select__item--hover {
  @include select__item--hover();
}
