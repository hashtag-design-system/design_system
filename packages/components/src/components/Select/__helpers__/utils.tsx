import { render, RenderOptions } from "@testing-library/react";
import React from "react";
import { SelectContextProvider, SelectContextType } from "../../../utils";

export const defaultProps: Pick<
  SelectContextType,
  "isOpen" | "isMobile" | "multiSelectable" | "items" | "value" | "width" | "filterValue"
> = {
  isOpen: false,
  isMobile: false,
  multiSelectable: false,
  value: "",
  items: [],
  width: "200px",
  filterValue: [],
};

type SelectCustomRenderOptions = {
  providerProps?: Partial<SelectContextType>;
  renderOptions?: RenderOptions;
};

export const selectCustomRender = (ui: React.ReactElement, options?: SelectCustomRenderOptions) => {
  return render(
    // @ts-expect-error
    <SelectContextProvider value={{ ...defaultProps, ...options?.providerProps }}>{ui}</SelectContextProvider>,
    options?.renderOptions
  );
};
