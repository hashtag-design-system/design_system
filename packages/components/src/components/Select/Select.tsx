import React, { useEffect, useRef, useState } from "react";
import { SelectContextProvider } from "../../utils/contexts";
import { useClassnames, useClickOutside, useIsMobile } from "../../utils/hooks";
import { ComponentProps } from "../__helpers__";
import { Button } from "./Button";
import { Filter } from "./Filter";
import { Header } from "./Header";
import { Hr } from "./Hr";
import { Item } from "./Item";
import { Modal } from "./Modal";
import { Options } from "./Options";

export type SelectedItems = { id: string; content: string; highlightedChildren: string; selected: boolean; isShown: boolean };

export type Props = {
  defaultOpen?: boolean;
  multiSelectable?: boolean;
  mobileView?: boolean;
  onSelect?: (selected: SelectedItems[]) => void;
  onDismiss?: (e: React.MouseEvent<HTMLElement>) => void;
  onOutsideClick?: (outsideClick: boolean) => void;
};

type SubComponents = {
  Item: typeof Item;
  Header: typeof Header;
  Modal: typeof Modal;
  Button: typeof Button;
  Hr: typeof Hr;
  Options: typeof Options;
  Filter: typeof Filter;
};

export type FProps = Props &
  Omit<ComponentProps<"details", false>, "onSelect"> & {
    forwardRef?: ComponentProps<"details", true>["ref"];
  };

export type SBProps = Props & Pick<FProps, "defaultOpen" | "onToggle" | "onSelect">;

const Select: React.FC<FProps> & SubComponents = ({
  defaultOpen = false,
  multiSelectable = false,
  mobileView = false,
  children,
  open,
  forwardRef,
  onToggle,
  onSelect,
  onOutsideClick,
  onDismiss,
  ...props
}) => {
  const { ref: modalRef, isOpen, setIsOpen, outsideClick } = useClickOutside<HTMLDivElement>(defaultOpen, undefined, onDismiss);
  const [value, setValue] = useState<string>("");
  const [items, setItems] = useState<SelectedItems[]>([]);
  const [isDisabled, setIsDisabled] = useState(false);
  const [classNames, rest] = useClassnames("select__box__container", props);
  const { isMobile } = useIsMobile(mobileView);

  useEffect(() => {
    if (onOutsideClick) {
      onOutsideClick(outsideClick);
    }
  }, [outsideClick, onOutsideClick]);

  const divRef = useRef<HTMLDivElement>(null);

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (isDisabled) {
      e.preventDefault();
    }
  };

  const handleToggle = (e: React.SyntheticEvent<HTMLElement>, boolean = true) => {
    e.preventDefault();

    if (!isDisabled) {
      if (boolean) {
        // @ts-expect-error
        setIsOpen(e.target.open);
      } else {
        setIsOpen(!isOpen);
      }
    }

    if (onToggle) {
      onToggle(e);
    }
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.target instanceof HTMLInputElement) {
      return;
    }
    e.preventDefault();

    const focusItem = (next: boolean) => {
      if (divRef && divRef.current) {
        const options = Array.from(divRef.current.getElementsByClassName("select__item"));
        const selected = document.activeElement;

        let item: Element = options[0];
        if (selected && selected.tagName.toLowerCase() === "div") {
          const idx = options.indexOf(selected);
          if (next) {
            const nextOpt = options[idx + 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            }
          } else {
            const nextOpt = options[idx - 1];
            if (nextOpt) {
              item = nextOpt;
              // (nextOpt as HTMLDivElement).focus();
            } else {
              item = options[options.length - 1];
            }
          }
        }
        return item;
      }
    };

    const isSummaryFocused = e.target instanceof Element && e.target.tagName.toLowerCase() === "summary";
    switch (e.code) {
      case "Escape":
        if (isOpen) {
          handleToggle(e, false);
        }
        break;
      case "Space":
      case "Enter":
        handleToggle(e, false);

        const selected = document.activeElement;
        if (!isSummaryFocused) {
          if (selected && selected.className === "select__item") {
            // @ts-expect-error
            document.activeElement.click();
            e.preventDefault();
          }
        }

        break;
      case "ArrowDown":
        if (isSummaryFocused && !isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(true);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
      case "ArrowUp":
        if (isSummaryFocused && isOpen) {
          handleToggle(e, false);
        } else {
          const target = focusItem(false);
          if (target) {
            (target as HTMLDivElement).focus();
          }
          e.preventDefault();
        }
        break;
    }
  };

  useEffect(() => {
    setValue(
      items
        .filter(item => item.selected)
        .map(item => item.content)
        .join(", ")
    );

    if (onSelect && items.length >= 1) {
      onSelect(items);
    }
  }, [items, onSelect]);

  const fOpen = open === undefined ? isOpen : open;

  return (
    <SelectContextProvider
      value={{
        isOpen: fOpen,
        ref: forwardRef,
        value,
        multiSelectable,
        items,
        isMobile,
        modalRef,
        setItems,
        handleToggle,
        setIsDisabled,
      }}
    >
      <div className="select__container" ref={divRef} data-testid="select-container">
        <details
          ref={isMobile ? undefined : modalRef}
          className={classNames}
          open={fOpen}
          onClick={e => handleClick(e)}
          onToggle={e => handleToggle(e)}
          onKeyDown={e => handleKeyDown(e)}
          data-testid="select"
          {...rest}
        >
          {children}
        </details>
      </div>
    </SelectContextProvider>
  );
};

Select.Header = Header;
Select.Item = Item;
Select.Modal = Modal;
Select.Button = Button;
Select.Hr = Hr;
Select.Options = Options;
Select.Filter = Filter;

export default Select;
