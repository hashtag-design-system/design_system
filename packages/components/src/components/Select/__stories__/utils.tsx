import Select from "../index";

export type StoriesComponentProps = { headerChildren?: boolean; optionsChilden?: boolean };

export const StoriesItems: React.FC<Pick<StoriesComponentProps, "optionsChilden">> = ({ optionsChilden = false }) => {
  return (
    <>
      {!optionsChilden && <Select.Filter placeholder="Filter" floatingplaceholder={false} />}
      <Select.Options>
        {optionsChilden && <Select.Filter placeholder="Filter" floatingplaceholder={false} />}
        <Select.Item id="hey_george">Hey george</Select.Item>
        <Select.Item id="amsterdam">
          Amsterdam<p>george</p>
        </Select.Item>
        <Select.Item id="georgekrax">georgekrax</Select.Item>
        <Select.Item id="georgekrax2">georgekrachropoulos</Select.Item>
        <Select.Item id="test">Test item</Select.Item>
        <Select.Item id="test_id">
          <strong>NL</strong> Amsterdam
        </Select.Item>
      </Select.Options>
    </>
  );
};

export const StoriesComponent: React.FC<any & StoriesComponentProps> = ({
  headerChildren = false,
  optionsChilden = false,
  align = "left",
  ...props
}) => {
  return (
    <Select {...props}>
      <Select.Button style={{ width: "200px" }}>Projects</Select.Button>
      <Select.Modal align={align}>
        {headerChildren ? (
          <Select.Header value="header" {...props}>
            <StoriesItems optionsChilden={optionsChilden} />
          </Select.Header>
        ) : (
          <>
            <Select.Header value="header" {...props} />
            <StoriesItems optionsChilden={optionsChilden} />
          </>
        )}
      </Select.Modal>
    </Select>
  );
};
