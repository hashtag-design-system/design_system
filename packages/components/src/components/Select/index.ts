import Select from "./Select";

export type { FProps as SelectButtonFProps, Props as SelectButtonProps } from "./Button";
export type { FProps as SelectFilterFProps, Props as SelectFilterProps } from "./Filter";
export type { FProps as SelectHeaderFProps } from "./Header";
export type { FProps as SelectHrFProps } from "./Hr";
export type { FProps as SelectItemFProps } from "./Item";
export type { FProps as SelectModalFProps, Props as SelectModalProps } from "./Modal";
export type { FProps as SelectOptionsFProps } from "./Options";
export type { FProps as SelectFProps, Props as SelectProps } from "./Select";

export default Select;
