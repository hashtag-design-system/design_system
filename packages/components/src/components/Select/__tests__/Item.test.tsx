import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SelectContextType } from "../../../utils/contexts";
import Select from "../index";
import { selectCustomRender } from "../__helpers__/utils";

const TEST_SELECTED_MULTIPLE_ITEMS: SelectContextType["items"] = [
  { id: "test_id0", content: "Test 1", selected: true, isShown: true },
  { id: "test_id1", content: "Test 2", selected: false, isShown: false },
];

describe("<Select.Item />", () => {
  test("default behaviour", () => {
    render(
      <Select defaultOpen>
        <Select.Item id="test_id" />
      </Select>
    );
    const item = screen.getByTestId("select-item");
    const itemInput = screen.getByTestId("select-item-input");
    const itemLabel = screen.getByTestId("select-item-label");

    // item tests
    expect(item).toBeVisible();
    expect(item.onclick).toBeDefined();
    expect(item.tagName.toLowerCase()).toBe("div");
    expect(item).toHaveAttribute("class");
    expect(item).toHaveAttribute("tabindex", "0");
    expect(item).toHaveAttribute("aria-selected", "false");
    expect(item.children).toHaveLength(2);
    expect(item.onclick).toBeDefined();

    // itemInput tests
    expect(itemInput).toBeVisible();
    expect(itemInput.tagName.toLowerCase()).toBe("input");
    expect(itemInput).toHaveAttribute("class");
    expect(itemInput).toHaveAttribute("id", "test_id");
    expect(itemInput).toHaveAttribute("value", "false");
    expect(itemInput).toHaveAttribute("type", "checkbox");
    expect(itemInput).toHaveAttribute("aria-checked", "false");

    // itemLabel tests
    expect(itemLabel).toBeVisible();
    expect(itemLabel).toHaveAttribute("class");
    expect(itemLabel).toHaveAttribute("for", "test_id");
    expect(itemLabel.children).toHaveLength(0);
    expect(itemLabel.textContent).toBe("");
  });
  test("with defaultChecked={true}", () => {
    render(
      <Select defaultOpen>
        <Select.Item id="test_id" defaultChecked />
      </Select>
    );
    const item = screen.getByTestId("select-item");
    const itemInput = screen.getByTestId("select-item-input");

    expect(item).toHaveAttribute("aria-selected", "true");
    expect(item).toMatchSnapshot();

    expect(itemInput).toHaveAttribute("value", "true");
    expect(itemInput).toHaveAttribute("type", "checkbox");
    expect(itemInput).toHaveAttribute("aria-checked", "true");
  });
  test('state="disabled"', () => {
    render(
      <Select defaultOpen>
        <Select.Button style={{ width: "200px" }}>Project</Select.Button>
        <Select.Item id="test_id" state="disabled">
          Test
        </Select.Item>
      </Select>
    );
    const item = screen.getByTestId("select-item");
    const itemInput = screen.getByTestId("select-item-input");

    expect(item).toHaveClass("disabled");
    expect(item).toHaveAttribute("aria-selected", "false");
    expect(item).toHaveAttribute("aria-disabled", "true");
    expect(item).toMatchSnapshot();

    userEvent.click(item);

    expect(itemInput).toBeDisabled();
    expect(itemInput).toHaveAttribute("value", "false");
    expect(itemInput).toHaveAttribute("aria-checked", "false");
    expect(itemInput).toHaveAttribute("aria-disabled", "true");

    const btnTextContent = screen.getByTestId("select-btn").children[0].textContent;
    expect(btnTextContent).not.toBe("Test");
    expect(btnTextContent).toBe("Project");
  });
  test("with children", () => {
    render(
      <Select defaultOpen>
        <Select.Button style={{ width: "200px" }}>Project</Select.Button>
        <Select.Item id="test_id">
          <strong>NL</strong>Amsterdam
        </Select.Item>
      </Select>
    );
    const itemLabel = screen.getByTestId("select-item-label");
    const children = itemLabel.children;

    expect(itemLabel).toHaveTextContent("Amsterdam");
    expect(children).toHaveLength(1);
    expect(children[0].tagName.toLowerCase()).toBe("strong");
    expect(children[0].textContent).toBe("NL");

    userEvent.click(itemLabel);
    // Check for items["content"] serialization in <Select.Item /> (newChildren)
    expect(screen.getByTestId("select-btn").children[0]).toHaveTextContent("NL Amsterdam");
  });
  test("with mobileView={true}", () => {
    render(
      <Select defaultOpen mobileView>
        <Select.Item id="test_id" />
      </Select>
    );

    expect(screen.getByTestId("select-hr")).toBeVisible();
  });
  test("with disabled={true}", () => {
    render(
      <Select defaultOpen>
        <Select.Item id="test_id" aria-disabled="true" />
      </Select>
    );
    const item = screen.getByTestId("select-item");
    const selectInput = screen.getByTestId("select-item-input");

    expect(item).toHaveAttribute("tabindex", "-1");
    expect(item).toHaveAttribute("aria-disabled", "true");
    expect(selectInput).toBeDisabled();
    expect(selectInput).toHaveAttribute("aria-disabled", "true");
  });
  test("with isShown={false}", () => {
    selectCustomRender(<Select.Item id={TEST_SELECTED_MULTIPLE_ITEMS[1].id}>{TEST_SELECTED_MULTIPLE_ITEMS[1].content}</Select.Item>, {
      // Pass for document - screen render, otherwise it will not be displayed, so we won't have access to it
      providerProps: { items: TEST_SELECTED_MULTIPLE_ITEMS.slice(1, 2) },
    });

    expect(screen.queryAllByTestId("select-item")).toHaveLength(0);
  });
  test("with one selectedItems", () => {
    render(
      <Select defaultOpen>
        <Select.Item id={TEST_SELECTED_MULTIPLE_ITEMS[0].id} defaultChecked={TEST_SELECTED_MULTIPLE_ITEMS[0].selected}>
          {TEST_SELECTED_MULTIPLE_ITEMS[0].content}
        </Select.Item>
      </Select>
    );
    // , {
    //   // Pass for document - screen render, otherwise it will not be displayed, so we won't have access to it
    //   providerProps: { items: TEST_SELECTED_MULTIPLE_ITEMS.slice(0, 1) },
    // });
    const item = screen.getByTestId("select-item");
    const itemInput = screen.getByTestId("select-item-input");

    expect(item).toHaveAttribute("aria-selected", "true");

    expect(itemInput).toHaveAttribute("value", "true");
    expect(itemInput).toHaveAttribute("aria-checked", "true");
    expect(itemInput).toHaveAttribute("id", TEST_SELECTED_MULTIPLE_ITEMS[0].id);
  });
  test("with multiple selectedItems", async () => {
    render(
      <Select defaultOpen>
        {/* This does not work in testing env: {TEST_SELECTED_MULTIPLE_ITEMS.map(({ id, content }) => {
          <Select.Item id={id}>{content}</Select.Item>;
        })} */}
        <Select.Item id={TEST_SELECTED_MULTIPLE_ITEMS[0].id} defaultChecked={TEST_SELECTED_MULTIPLE_ITEMS[0].selected}>
          {TEST_SELECTED_MULTIPLE_ITEMS[0].content}
        </Select.Item>
        <Select.Item id={TEST_SELECTED_MULTIPLE_ITEMS[1].id} defaultChecked={TEST_SELECTED_MULTIPLE_ITEMS[1].selected}>
          {TEST_SELECTED_MULTIPLE_ITEMS[1].content}
        </Select.Item>
      </Select>
    );

    screen.getAllByTestId("select-item").forEach((item, i) => {
      const checked = TEST_SELECTED_MULTIPLE_ITEMS[i].selected;
      expect(item.getAttribute("aria-selected")).toBe(checked.toString());
    });

    screen.getAllByTestId("select-item-input").forEach((input, i) => {
      const checked = TEST_SELECTED_MULTIPLE_ITEMS[i].selected;
      expect(input).toHaveAttribute("value", checked.toString());
      expect(input).toHaveAttribute("aria-checked", checked.toString());
      expect(input).toHaveAttribute("id", `test_id${i}`);
    });
  });
});
