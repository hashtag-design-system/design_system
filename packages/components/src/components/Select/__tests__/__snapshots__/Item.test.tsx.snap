// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select.Item /> with defaultChecked={true} 1`] = `
<div
  aria-selected="true"
  class="select__item"
  data-testid="select-item"
  tabindex="0"
>
  <input
    aria-checked="true"
    class="select__item__input"
    data-testid="select-item-input"
    id="test_id"
    type="checkbox"
    value="true"
  />
  <label
    class="body-14"
    data-testid="select-item-label"
    for="test_id"
    unselectable="on"
  />
</div>
`;
